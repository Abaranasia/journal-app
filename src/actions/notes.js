import { db } from "../firebase/firebaseConfig";
import { types } from "../types/types";

import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";


export const startNewNote = () => { // Acción para crear una nueva nota
    return async (dispatch, getState) => { //comunicación asíncrona con Thunk 

        const uid = getState().auth.uid;
        //console.log(uid)

        const newNote = {
            title: 'Otra nueva nota',
            body: 'jujijaje',
            date: new Date().getTime()
        }

        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote); // referencia al documento en Firestore donde añadir una nueva nota para este usuario
        // console.log(docRef)
        dispatch(activeNote(docRef.id, newNote))
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }

})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => { // acción asíncrona porque depende del backend

        const notes = await loadNotes(uid); // obtenemos todas las notas del usuario identificado y las enviaremos al store
        dispatch(setNotes(notes)) //hacemos el dispatch para enviar el listado de notas al store mediante la acción setNotes
    }
}


export const setNotes = (notes) => ({ // Acción para enviar al store las notas recibidas de Firebase
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => { // Acción para guardar una nota en la BBDD
    // necesitaremos el middleware Thunk para poder comunicarnos con Firebase
    return async (dispatch, getState) => { //Thunk, que es asíncrono, devuelve el dispatch y el uid del usuario

        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url; //Para evitar que Firebase pete si le llega un undefined
        }

        const noteToFirestore = { ...note };
        delete noteToFirestore.id; // Hacemos esto para extraer la id de note, dado que no la necesitamos en la inserción

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore); // Podemos enviar noteToFirestore o note, pro la primera no lleva id, ojo

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved!', note.title, 'success');
    }
}

export const refreshNote = (id, note) => ({ // Acción que permite refrescar una nota, en caso de que haya cambiado, por ejemplo, tras guardarla
    type: types.notesUpdated,
    payload: {
        id,
        note: {  // Tenemos que pasarlo así porque este note nos llega sin id al enviar noteToFirestore, asi que lo volvemos a poner
            id, //Sin este este id, jornalEntry nos da un error porque debe dibujar un elemento sin key
            ...note
        }

    }
})

