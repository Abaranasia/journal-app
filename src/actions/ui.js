import { types } from '../types/types';

/** Este mÃ³dulo define los actions en referencia a los error del sistema */

export const setError = (err) => ({
    type: types.uiSetError,
    payload: err
})

export const removeError = () => ({
    type: types.uiRemoveError
})

export const startLoading = () => ({
    type: types.uiStartLoading,

})

export const finishLoading = () => ({
    type: types.uiFinishLoading,

})