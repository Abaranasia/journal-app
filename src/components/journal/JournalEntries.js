import React from 'react'
import { useSelector } from 'react-redux';
import { JournayEntry } from './JournayEntry';

export const JournalEntries = () => {

    const { notes } = useSelector(state => state.notes); // obtenemos del store el listado de notes del usuario

    //console.log(notes);

    //const entries = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

    return (
        <div className="journal__entries ">

            {
                notes.map(note => (
                    <JournayEntry
                        key={note.id}
                        {...note}
                    />
                ))
            }

        </div>
    )
}
