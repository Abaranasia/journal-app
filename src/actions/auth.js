import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import { notesLogout } from "./notes";


/** Este módulo define los actions para el proceso de auth.
 *  Aquí incluiríamos todos los métodos de acceso necesarios: email, google, twitter, etc.
 */


export const startLoginEmailPassword = (email, password) => {
    // middleware para login con email que gestiona el action de login

    return (dispatch) => { //devuelve un objeto llamado dispatch con el tipo y el payload

        dispatch(startLoading()); // Empieza el procedimiento de login

        
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
            })
            .catch(e => {
                console.log('Hubo un error en el login:', e);
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
                Swal.fire('Error', e.message, 'error')
            });
/*         setTimeout(() => {
            dispatch(login(123, 'Pedro'))
        }, 3500); */

    }
}

export const startRegisterWithEmailPassword = (email, password, name) => {
    // dado que es ua tarea asíncrona, necesito retornar un callback
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name }); // hacemos esto porque el registro en Firebase no lo hace solo
                console.log(user);

                dispatch(
                    login(user.uid, user.displayName)
                )
            })
            .catch(e => { console.log('Hubo un error al registrarse:', e) });
    }
}

export const startGoogleLogin = () => {
    // middleware para login con usuario de Google que gestiona el action de login

    return (dispatch) => {
    //devuelve una función asíncrona que recibe un dispatch con el tipo y el payload que deben ser enviados

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                //console.log(user.uid, user.displayName);
                dispatch(
                    login(user.uid, user.displayName)
                )
            });
    }
}


export const login = (uid, displayName) => {
    //action para login: devuelve un tipo (login) y un payload con uid y nombre

    return {
        type: types.login,
        payload: {
            uid, displayName
        }
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout())
        dispatch(notesLogout())
    }
}
export const logout = (params) => ({
    type: types.logout
})
