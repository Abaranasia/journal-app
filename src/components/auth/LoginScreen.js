import React from 'react';
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";

import { useForm } from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';

export const LoginScreen = () => {

    /** Este módulo tiene acceso a useDispatch e invoca, dependiendo de la vía de acceso,
     * al action correspondiente para cada forma de acceso, ya que alguno requiere middleware */


    const dispatch = useDispatch(); // gestiona el dispatch de redux

    const { loading } = useSelector(state => state.ui) // Esto nos permite identificar si cuándo se está haciendo login

    const [formValues, handleInputChange] = useForm({ //useForm
        email: 'rankirlian@gmail.com',
        password: 'abc123'
    });

    const { email, password } = formValues;


    const handleLogin = (e) => { // para login con email y password
        e.preventDefault();
        //console.log(email, password);
        //dispatch(login(1028201, 'Ran'))
        dispatch(startLoginEmailPassword(email, password)) // recibe type y payload para enviar al dispatcher
    }


    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin()); // recibe type y payload para enviar al dispatcher
    }


    return (
        <>
            <h3 className="auth__title">Login</h3>

            <form onSubmit={handleLogin} className="animate__animated animate__fadeIn animate__faster">
                <input
                    type="text"
                    placeholder="email"
                    name="email"
                    value={email}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="password"
                    name="password"
                    value={password}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary"
                    disabled={loading}
                >
                    Login
                </button>

                <br />
                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link to="/auth/register">
                    Create new account
                </Link>
            </form>
        </>
    )
}
