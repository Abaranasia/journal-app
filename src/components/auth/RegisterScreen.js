import React from 'react';
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from 'react-redux';

import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPassword } from '../../actions/auth';

export const RegisterScreen = () => {

    const dispatch = useDispatch();

    const { msgError } = useSelector(state => state.ui);
    /** El hook useSelecto permite redirigir info desde el Store a la app
     * Aquí recibe el state de store y lo leemos para poder usar el mensaje de error */
   // console.log(msgError)

    const [formValues, handleInputChange] = useForm({
        name: 'Ran Kirlian',
        email: 'rankirlian@gmail.com',
        password: 'abc123',
        password2: 'abc123'
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            //console.log(name, email, password, password2);
            dispatch(startRegisterWithEmailPassword(email, password, name))
        }
    }

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            //console.log('Name is required');
            return false;
        } //else if (email... contra validator.js) {}
        else if (email.trim().length === 0) {
            dispatch(setError('email is required'));
            //console.log('email is required');
            return false;
        } else if (password !== password2 || password.length < 3) {
            dispatch(setError('password missmatch or malformed'));
            //console.log('password missmatch or malformed');
            return false;
        }
        dispatch(removeError())
        return true
    }


    return (
        <>
            <h3 className="auth__title">Register</h3>

            <form onSubmit={handleRegister} className="animate__animated animate__fadeIn animate__faster">

                {msgError && // Pinta la capa si hay un mensaje de error y en él muestra dicho error
                    (<div className="auth__alert-error">
                        {msgError}
                    </div>)
                }

                <input
                    type="text"
                    placeholder="name"
                    name="name"
                    value={name}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="text"
                    placeholder="email"
                    name="email"
                    value={email}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="password"
                    name="password"
                    value={password}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="confirm password"
                    name="password2"
                    value={password2}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary mb-5"
                >
                    Register
                </button>
                <br />
                <Link to="/auth/login">
                    Already registered?
                </Link>
            </form>
        </>
    )
}