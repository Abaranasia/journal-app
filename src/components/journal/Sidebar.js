import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { JournalEntries } from './JournalEntries';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';


export const Sidebar = () => {


    const dispatch = useDispatch();
    const { name } = useSelector(state => state.auth) // Esto nos permite extrar desde el state el nombre del usuario conectado

    const handleLogout = () => {
        dispatch(startLogout()) //invoca al action de logout en auth.js

    }

    const handleAddNew = () => {
        dispatch(startNewNote());

    }


    return (
        <aside className="journal__sidebar">

            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    <i className="far fa-moon"></i>
                    {name}
                </h3>
                <button className="btn" onClick={handleLogout}>Lougout</button>
            </div>

            <div className="journal__new-entry" onClick={handleAddNew}>
                <i className="far fa-calendar-plus fa-3x"></i>
                <p className="mt-5">New entry</p>
            </div>

            <JournalEntries />
        </aside>
    )
}
