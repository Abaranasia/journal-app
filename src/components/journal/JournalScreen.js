import React from 'react';
import { useSelector } from 'react-redux';
import { NoteScreen } from '../notes/NoteScreen';
import { NothingSelected } from './NothingSelected';
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {

    const { active: activeNote } = useSelector(state => state.notes); //para saber desde el reducer si la nota est√° o no activa

    return (
        <div className="journal__main-content animate__animated animate__fadeIn animate__faster">
            <Sidebar />

            <main>
                {
                    (activeNote)
                        ? (<NoteScreen />)
                        : (<NothingSelected />)
                }

            </main>
        </div>
    )
}
