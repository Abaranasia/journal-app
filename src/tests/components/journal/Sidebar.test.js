import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
//import { firebase } from '../../firebase/firebaseConfig';

import { Sidebar } from '../../../components/journal/Sidebar';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';

//import { getAuth } from 'firebase/auth';
//const auth = getAuth();

//Evita el crasheo de tipo: INTERNAL ASSERTION FAILED: Expected a class definition en  firebase.auth()


jest.mock('../../../actions/auth', () => ({
    // Mockeamos estas funciones porque ya están probadas y no necesitamos volver a hacerlo aquí
    startLogout: jest.fn(),
})
);

jest.mock('../../../actions/notes', () => ({
    // Mockeamos estas funciones porque ya están probadas y no necesitamos volver a hacerlo aquí
    startNewNote: jest.fn(),
})
);

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Para evaluar estas pruebas vamos a necesitar que nos ayude el store. Para ello empleamos mockStore, 
// MockStore nos permitirá averiguarlas acciones que lo han disparado, con qué argumentos, etc.

const initState = { //Debe tener el estado inicial, el que nos encontraremos antes de hacer login 
    auth: {
        udi: "1",
        name: "Ran"
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: null,
        notes: []
    }
}

let store = mockStore(initState);
// Nuestro store fictício debe definir aquí el estado inicial en el que necesitamos que esté para estos test
store.dispatch = jest.fn(); // Mokeamos el dispatch porque necesitamos evaluarlo aquí


const wrapper = mount(
    <Provider store={store}> {/* lo necesitamos para fingir el context del usuario */}
        <Sidebar />
    </Provider>
)

describe('<Sidebar /> Tests', () => {


    test('should render properly', () => {
        expect(wrapper).toMatchSnapshot()
    });



    test('should call logout action on click', () => {
        wrapper.find('button').prop('onClick')();

        expect(startLogout).toHaveBeenCalled()
    });



    test('should call startnewNote on click', () => {
        wrapper.find('.journal__new-entry').prop('onClick')();

        expect(startNewNote).toHaveBeenCalled()
    });
});