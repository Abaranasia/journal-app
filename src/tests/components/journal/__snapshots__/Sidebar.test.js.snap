// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Tests should render properly 1`] = `
Array [
  " ",
  <aside
    className="journal__sidebar"
  >
    <div
      className="journal__sidebar-navbar"
    >
      <h3
        className="mt-5"
      >
        <i
          className="far fa-moon"
        />
        Ran
      </h3>
      <button
        className="btn"
        onClick={[Function]}
      >
        Lougout
      </button>
    </div>
    <div
      className="journal__new-entry"
      onClick={[Function]}
    >
      <i
        className="far fa-calendar-plus fa-3x"
      />
      <p
        className="mt-5"
      >
        New entry
      </p>
    </div>
    <div
      className="journal__entries "
    />
  </aside>,
]
`;
